kernel/os/src/os_sem.c

```diff
@@ -42,6 +42,7 @@ os_sem_init(struct os_sem *sem, uint16_t tokens)
         goto done;
     }
 
+    sem->max_tokens = tokens;
```

Comment:
`tokens` parameter is an initial value, not a max value
we can only check for an integer overflow on `sem_tokens`

Reply:
@andrzej-kaczmarek agreed, but wouldn't the problem also be to pass the initial value? probably more likely to happen then to OF an uint16_t.

Reply:
it's ok to pass any initial value, it depends on the purpose of semaphore.
and as you said, it's unlikely to overflow a 16-bit counter in a semaphore so most likely this is why we do not care about it :-)

Reply:
If the limit was a uint8_t we would have checked for token overflow. If you are overflowing a 16-bit token for a semaphore you are not using it correctly :-) Or there is something very, very wrong. I would even consider asserting if the tokens would overflow as it simply should never happen.

Reply:
Ok, so I'll probably drop this

Reply:
I would think that `tokens` is the number of resources I want to protect. Let's say I can allow 4 devices access some resource (I might have 4 caches, etc), and then one after one 4 of them  `_pend` which does not allow any more to use until some of them `_release`. But say for some reason (bug ;-) someone that had a lock releases it two times. Now we have a counter that allow 5 consumers which probably means a data race. This could happen with open/closing devices, etc. So this is not how the API is supposed to be?

Reply:
@utzig imposing a limit on semaphore count based on initial value would make it useless as a signal (which is most common semaphore usage in core repo right now)